### PR TITLE
feature/jarm_sdk

### PR DESCRIPTION
**Pull request by sanderPostma**
Fixes while testing jarm

### CODE REVIEW

File: packages/did-auth-siop-adapter/lib/did/DidJWT.ts [hunk 1]
```diff
@@ -238,7 +238,7 @@ function assertIssSelfIssuedOrDid(payload: JWTPayload) {
   }
 }
 
-export function getSubDidFromPayload(payload: JWTPayload, header?: JWTHeader): string {
+export function getSubDidFromPayload(payload: JWTPayload, header?: JWTHeader): string | undefined {
```
Comment (nklomp) [inline packages/did-auth-siop-adapter/lib/did/DidJWT.ts:241]: Please revert. The assert method below ensures a string is always present

File: packages/siop-oid4vp/lib/__tests__/AuthenticationRequest.request.spec.ts [hunk 2]
```diff
@@ -98,6 +98,26 @@ describe('create Request Uri should', () => {
         response_type: 'id_token',
         request_object_signing_alg_values_supported: [SigningAlgo.EDDSA, SigningAlgo.ES256],
         redirect_uri: EXAMPLE_REDIRECT_URL,
+        claims: {
```
Comment (nklomp) [inline packages/siop-oid4vp/lib/__tests__/AuthenticationRequest.request.spec.ts:101]: Why do we need this all of a sudden? claims are only required when it really is OID4VP. Not for SIOPv2

File: packages/siop-oid4vp/lib/__tests__/AuthenticationRequest.request.spec.ts [hunk 3]
```diff
@@ -268,10 +362,10 @@ describe('create Request Uri should', () => {
     const uriRequest = await URI.fromOpts(opts)
 
     const uriDecoded = decodeURIComponent(uriRequest.encodedUri)
-    expect(uriDecoded).toContain(`openid://?request=eyJhbGciOi`)
+    expect(uriDecoded).toContain(`openid4vp://?client_id=https://www.example.com/`)
```
Comment (nklomp) [inline packages/siop-oid4vp/lib/__tests__/AuthenticationRequest.request.spec.ts:365]: Please do not remove the test for request = jwt. This obviously should have been a new expect. And then not toContain but a check that it starts with

File: packages/siop-oid4vp/lib/__tests__/AuthenticationRequest.request.spec.ts [hunk 4]
```diff
@@ -511,30 +605,33 @@ describe('create Request JWT should', () => {
   it('succeed when requesting with a valid PD', async () => {
     const opts: CreateAuthorizationRequestOpts = {
       version: SupportedVersion.SIOPv2_ID1,
-      /*payload: {
+      payload: {
```
Comment (nklomp) [inline packages/siop-oid4vp/lib/__tests__/AuthenticationRequest.request.spec.ts:608]: Why is the payload fully re-enabled, whilst this test already contains a request object. If you need to add things add them there

File: packages/siop-oid4vp/lib/authorization-response/OpenID4VP.ts [hunk 5]
```diff
@@ -278,21 +283,25 @@ export const assertValidVerifiablePresentations = async (args: {
...
   PresentationExchange.assertValidPresentationDefinitionWithLocations(args.presentationDefinitions)
-  const presentationsWithFormat = args.presentations
 
   if (
     args.presentationDefinitions &&
     args.presentationDefinitions.length &&
     (!presentationsWithFormat || (Array.isArray(presentationsWithFormat) && presentationsWithFormat.length === 0))
   ) {
-    throw new Error(SIOPErrors.AUTH_REQUEST_EXPECTS_VP)
+    if(nrOfMdocsPresentations === 0) {
```
Comment (nklomp) [inline packages/siop-oid4vp/lib/authorization-response/OpenID4VP.ts:302]: Whut? At least make a comment on why something like this would be needed

File: packages/siop-oid4vp/lib/authorization-response/PresentationExchange.ts [hunk 6]
```diff
@@ -76,8 +82,11 @@ export class PresentationExchange {
         keyEncoding: options?.signatureOptions?.keyEncoding ?? KeyEncoding.Hex,
       },
     }
+    return await this.pex.verifiablePresentationFrom(presentationDefinition, this.removeMDocCredentials(selectedCredentials), presentationSignCallback, signOptions)
```
Comment (nklomp) [inline packages/siop-oid4vp/lib/authorization-response/PresentationExchange.ts:85]: Please make comments when you make these types of exceptions as to why it is needed

File: packages/siop-oid4vp/lib/authorization-response/PresentationExchange.ts [hunk 7]
```diff
@@ -104,7 +113,7 @@ export class PresentationExchange {
     } else if (!this.allVerifiableCredentials || this.allVerifiableCredentials.length == 0) {
       throw new Error(`${SIOPErrors.COULD_NOT_FIND_VCS_MATCHING_PD}, no VCs were provided`)
     }
-    const selectResults: SelectResults = this.pex.selectFrom(presentationDefinition, this.allVerifiableCredentials, {
+    const selectResults: SelectResults = this.pex.selectFrom(presentationDefinition, this.removeMDocCredentials(this.allVerifiableCredentials), {
```
Comment (nklomp) [inline packages/siop-oid4vp/lib/authorization-response/PresentationExchange.ts:116]: Please make comments if you make these types of exceptions as to why it is needed

File: packages/siop-oid4vp/lib/authorization-response/PresentationExchange.ts [hunk 8]
```diff
@@ -324,7 +383,7 @@ export class PresentationExchange {
       throw new Error(SIOPErrors.NO_PRESENTATION_SUBMISSION)
     }
 
-    if (evaluationResults.areRequiredCredentialsPresent === Status.ERROR || evaluationResults.errors?.length || !evaluationResults.value) {
+    if (!evaluationResults.areRequiredCredentialsPresent || (evaluationResults.errors && evaluationResults.errors.length > 0) || !evaluationResults.value) {
```
Comment (nklomp) [inline packages/siop-oid4vp/lib/authorization-response/PresentationExchange.ts:386]: The first part of the expression all of a sudden does something very different from the previous version, given it is an enum

File: packages/siop-oid4vp/lib/helpers/Metadata.ts [hunk 9]
```diff
@@ -87,26 +87,41 @@ function supportedSubjectSyntaxTypes(rpMethods: string[] | string, opMethods: st
   return supportedSubjectSyntaxTypes
 }
 
+export function collectAlgValues(o: any): string[] {
```
Comment (nklomp) [inline packages/siop-oid4vp/lib/helpers/Metadata.ts:90]: Do not name an arg `o` on a public method

File: packages/siop-oid4vp/lib/helpers/Metadata.ts [hunk 10]
```diff
@@ -87,26 +87,41 @@ function supportedSubjectSyntaxTypes(rpMethods: string[] | string, opMethods: st
...
+    }
+    if (opFormatElement !== undefined) {
+      Object.keys(opFormatElement).forEach((k) => (methodKeyOP = k));
+    }
     if (methodKeyRP !== methodKeyOP) {
       throw new Error(SIOPErrors.CREDENTIAL_FORMATS_NOT_SUPPORTED)
     }
     const algs = getIntersection(rpAlgs, opAlgs)
-    if (!algs.length) {
+    if (!algs.length && isJwtFormat(crFormat)) {
```
Comment (nklomp) [inline packages/siop-oid4vp/lib/helpers/Metadata.ts:124]: Huh. RP and OP support certain algorithms. What does that have to do with JWT only?

File: packages/siop-oid4vp/lib/helpers/Revocation.ts [hunk 11]
```diff
@@ -10,25 +16,27 @@ export const verifyRevocation = async (
   if (!vpToken) {
     throw new Error(`VP token not provided`)
   }
-  if (!revocationVerificationCallback) {
-    throw new Error(`Revocation callback not provided`)
-  }
+  if (isWrappedW3CVerifiablePresentation(vpToken) || isWrappedSdJwtVerifiablePresentation(vpToken)) { // Only W3C support for now
```
Comment (nklomp) [inline packages/siop-oid4vp/lib/helpers/Revocation.ts:19]: Please invert that and create a debug line stating it is not supported for non-w3c and return immediately, instead of creating a big if

File: packages/siop-oid4vp/lib/op/OPBuilder.ts [hunk 12]
```diff
@@ -13,7 +13,7 @@ export class OPBuilder {
   expiresIn?: number
   issuer?: IIssuerId | ResponseIss
   responseMode?: ResponseMode = ResponseMode.DIRECT_POST
-  responseRegistration?: Partial<ResponseRegistrationOpts> = {}
+  responseRegistration?: Partial<ResponseRegistrationOpts> //= {}
```
Comment (nklomp) [inline packages/siop-oid4vp/lib/op/OPBuilder.ts:16]: This is likely to cause issues in certain situations as the builder properties are assigned to the OP in the build phase. So why was this needed?

File: packages/siop-oid4vp/lib/rp/Opts.ts [hunk 13]
```diff
@@ -55,6 +60,7 @@ export const createVerifyResponseOptsFromBuilderOrExistingOpts = (opts: {
   return opts.builder
     ? {
         hasher: opts.builder.hasher ?? defaultHasher,
+//        correlationId: uuidv4(), We can't set a correlationId here, it will no longer check functions like  this.sessionManager.getCorrelationIdByNonce(resNonce, false)
```
Comment (nklomp) [inline packages/siop-oid4vp/lib/rp/Opts.ts:63]: Yes, why would it need that function to begin with if you already have a correlation Id? and internally we work by correlationId

File: packages/siop-oid4vp/lib/rp/RP.ts [hunk 14]
```diff
@@ -196,13 +198,13 @@ export class RP {
         correlationId,
       })
       correlationId = verifyAuthenticationResponseOpts.correlationId ?? correlationId
-      void this.emitEvent(AuthorizationEvents.ON_AUTH_RESPONSE_RECEIVED_SUCCESS, {
+      await this.emitEvent(AuthorizationEvents.ON_AUTH_RESPONSE_RECEIVED_SUCCESS, {
```
Comment (nklomp) [inline packages/siop-oid4vp/lib/rp/RP.ts:201]: First why is this needed. This makes the events sync. Second. If you change it here, then at least change it everywhere in this file

File: packages/siop-oid4vp/lib/rp/RP.ts [hunk 15]
```diff
@@ -225,6 +227,19 @@ export class RP {
     return this._verifyResponseOptions
   }
 
+  public getResponseRedirectUri(mappings?: Record<string, string>): string | undefined {
+    if (this._responseRedirectUri === undefined) {
```
Comment (nklomp) [inline packages/siop-oid4vp/lib/rp/RP.ts:231]: Why are you constantly explicitly checking against undefined. That can have undesired side effects if someone explicitly sets something to null for instance (which is a valid value for an optional const/var/param). Simply do `!param`

File: packages/siop-oid4vp/lib/rp/RP.ts [hunk 16]
```diff
@@ -225,6 +227,19 @@ export class RP {
...
+      return undefined
+    }
+    if(mappings === undefined) {
```
Comment (nklomp) [inline packages/siop-oid4vp/lib/rp/RP.ts:234]: Why are you constantly explicitly checking against undefined. That can have undesired side effects if someone explicitly sets something to null for instance (which is a valid value for an optional const/var/param). Simply do `!param`

**Review by nklomp — CHANGES_REQUESTED** [review]
Please see remarks. Mostly small

File: packages/siop-oid4vp/lib/__tests__/spec-compliance/jwtVCPresentationProfile.spec.ts [hunk 17]
```diff
@@ -172,7 +172,7 @@ describe('RP using test vectors', () => {
     ).toBeTruthy()
   })
 
-  it('should decode auth response', async () => {
+  it.skip('should decode auth response', async () => { // FIXME pex is too lenient ATM
```
[the remark this answers is not in the document]
Comment (sanderPostma) [reply packages/siop-oid4vp/lib/__tests__/spec-compliance/jwtVCPresentationProfile.spec.ts:175]: Reverted, should pass now after updating pex to 5.0.0-unstable.18
That was the credential.vp resolution I removed in pex today.

File: packages/siop-oid4vp/lib/authorization-request/AuthorizationRequest.ts [hunk 18]
```diff
@@ -119,7 +120,10 @@ export class AuthorizationRequest {
     let requestObjectPayload: RequestObjectPayload | undefined = undefined
 
     const jwt = await this.requestObjectJwt()
-    const parsedJwt = jwt ? parseJWT(jwt) : undefined
+    let parsedJwt = undefined
```
[the remark this answers is not in the document]
Comment (sanderPostma) [reply packages/siop-oid4vp/lib/authorization-request/AuthorizationRequest.ts:123]: reverted. No clue why I did this...

File: packages/siop-oid4vp/lib/authorization-request/AuthorizationRequest.ts [hunk 19]
```diff
@@ -164,7 +168,10 @@ export class AuthorizationRequest {
       )
       assertValidRPRegistrationMedataPayload(registrationMetadataPayload)
       // TODO: We need to do something with the metadata probably
+    } /*else { this makes test mattr.launchpad.spec.ts fail why was this check added?
```
[the remark this answers is not in the document]
Comment (sanderPostma) [reply packages/siop-oid4vp/lib/authorization-request/AuthorizationRequest.ts:171]: Ah that was to have the following discussion:
These test are failing
image
It just creates an AuthorizationRequest for OpenBadge where RequestObject is undefined so there also won't be a parsedJwt / requestObjectPayload. Next mergedPayload is expected to have client_metadata which it won't, failing the test / CI pipeline with "could not fetch registrationMetadataPayload due to missing payload key client_metadata"

So the question is: is that assertion realistic to enforce? Or so we need to expand the test so it actually provides client_metadata?

File: packages/siop-oid4vp/lib/authorization-request/AuthorizationRequest.ts [hunk 20]
```diff
@@ -256,7 +270,11 @@ export class AuthorizationRequest {
   }
 
   public async mergedPayloads(): Promise<RequestObjectPayload> {
-    return { ...this.payload, ...(this.requestObject && (await this.requestObject.getPayload())) }
+    const  requestObjectPayload = { ...this.payload, ...(this.requestObject && (await this.requestObject.getPayload())) }
+    if (requestObjectPayload.scope && typeof requestObjectPayload.scope !== 'string') { //  test mattr.launchpad.spec.ts does not supply a scope value
```
[the remark this answers is not in the document]
Comment (sanderPostma) [reply packages/siop-oid4vp/lib/authorization-request/AuthorizationRequest.ts:274]: The situation I found was 
image

Not sure why Sadjad placed it precisely there, maybe because mergedPayloads() is called in many nasty/embedded places.
But the field should be there
> scope: string // REQUIRED. As specified in Section 3.1.2 of [OpenID.Core].

anyway the mattr.launchpad.spec.ts does not supply a requestObject so we also won't have a scope either.
so again these tests should either be upgraded or disabled imo

File: packages/did-auth-siop-adapter/lib/did/DidJWT.ts (hunk 1 above)
Comment (sanderPostma) [reply packages/did-auth-siop-adapter/lib/did/DidJWT.ts:241]: I see it now. But my tsc does not look inside assertIssSelfIssuedOrDid, so have to return payload.sub! instead

File: packages/siop-oid4vp/lib/authorization-response/OpenID4VP.ts (hunk 5 above)
Comment (sanderPostma) [reply packages/siop-oid4vp/lib/authorization-response/OpenID4VP.ts:302]: I guess I just wanted to make assertValidVerifiablePresentations not crash when running mdocs in which case presentationsWithFormat will be empty. (But the presentationDefinitions are still asserted)
Will add comment

File: packages/siop-oid4vp/lib/__tests__/AuthenticationRequest.request.spec.ts (hunk 2 above)
Comment (sanderPostma) [reply packages/siop-oid4vp/lib/__tests__/AuthenticationRequest.request.spec.ts:101]: removed again and tests are passing. I remember tests failing at some point, but something else must have been fixed by now

File: packages/siop-oid4vp/lib/__tests__/AuthenticationRequest.request.spec.ts (hunk 3 above)
Comment (sanderPostma) [reply packages/siop-oid4vp/lib/__tests__/AuthenticationRequest.request.spec.ts]: fixed

File: packages/siop-oid4vp/lib/__tests__/AuthenticationRequest.request.spec.ts (hunk 4 above)
Comment (sanderPostma) [reply packages/siop-oid4vp/lib/__tests__/AuthenticationRequest.request.spec.ts:609]: reverted.
I guess I must have been testing with a broken pex or something

File: packages/siop-oid4vp/lib/authorization-response/PresentationExchange.ts (hunk 7 above)
Comment (sanderPostma) [reply packages/siop-oid4vp/lib/authorization-response/PresentationExchange.ts:116]: done

File: packages/siop-oid4vp/lib/authorization-response/PresentationExchange.ts (hunk 8 above)
Comment (sanderPostma) [reply packages/siop-oid4vp/lib/authorization-response/PresentationExchange.ts:386]: weird.
I suppose I meant to only tighten up the second part.

Comment (sanderPostma) [reply packages/siop-oid4vp/lib/authorization-response/PresentationExchange.ts:386]: (fixed)

File: packages/siop-oid4vp/lib/helpers/Metadata.ts (hunk 9 above)
Comment (sanderPostma) [reply packages/siop-oid4vp/lib/helpers/Metadata.ts:90]: fixed, also it doesn't need export. (probably a ctrl-alt-m thing)

File: packages/siop-oid4vp/lib/authorization-request/AuthorizationRequest.ts (hunk 19 above)
[the remark this answers is not in the document]
Comment (nklomp) [reply packages/siop-oid4vp/lib/authorization-request/AuthorizationRequest.ts]: Uh, that means the test is wrong or something else is going on. The else is an assertion and should be there. There really needs to be a client_metadata(_uri) (previously called registration).

I don't get why you assume that what is clearly an assertion should all of a sudden not be needed anymore

File: packages/siop-oid4vp/lib/authorization-request/AuthorizationRequest.ts (hunk 20 above)
[the remark this answers is not in the document]
Comment (nklomp) [reply packages/siop-oid4vp/lib/authorization-request/AuthorizationRequest.ts:271]: Well something must be off with the mattr test or the way we treat request objects then. The test has been there for a very long time and always passed.

The lib should be able to work without a request object (so in plain oauth2 mode)

[the remark this answers is not in the document]
Comment (nklomp) [reply packages/siop-oid4vp/lib/authorization-request/AuthorizationRequest.ts:271]: Also the string assertion only applies of there is a scope to begin with. If not then the error will never be thrown

File: packages/siop-oid4vp/lib/authorization-response/OpenID4VP.ts (hunk 5 above)
Comment (nklomp) [reply packages/siop-oid4vp/lib/authorization-response/OpenID4VP.ts]: Well I believe current approach is wrong in any case. This triggers if there is no presentation. But then we only provide an error if there are no mdoc presentations. That doesn't make sense to me. 

That is always true anyway probably given mdoc is being filtered out above.

File: packages/siop-oid4vp/lib/helpers/Metadata.ts (hunk 10 above)
Comment (sanderPostma) [reply packages/siop-oid4vp/lib/helpers/Metadata.ts:124]: was attepmting to restructure this function, it looks complex when you first encounter it.
crFormat also has LdpObject & DiObject formats which will not produce any algs, so it only errors when the format key has 'jwt' or 'mdoc' in it.

File: packages/siop-oid4vp/lib/op/OPBuilder.ts (hunk 12 above)
Comment (sanderPostma) [reply packages/siop-oid4vp/lib/op/OPBuilder.ts:16]: I probably did this when fighting sudden issues with LanguageTagUtils.
```
    const languageTaggedFields: Map<string, string> = LanguageTagUtils.getLanguageTaggedProperties(
      opts.builder.responseRegistration,
      languageTagEnabledFieldsNames,
    )
```
when opts.builder.responseRegistration is undefined it's not going to try to translate. When an empty object it's going to throw error because of missing fields.

They also pass in languageTagEnabledFieldsNames which was actually interpreted as languageTagRequiredFieldsNames. But I think I fixed that at a later point, so I will try to revert that.

File: packages/siop-oid4vp/lib/__tests__/AuthenticationRequest.request.spec.ts (hunk 4 above)
Comment (sanderPostma) [reply packages/siop-oid4vp/lib/__tests__/AuthenticationRequest.request.spec.ts:528]: actually test "should throw error if presentation definition object is not valid" fails without the payload
It expects 
> Presentation definition in the request claims is not valid

but receives

> Received message:   "referenceUri specified, but object to host there is not present"

      14 |     if (opts.reference_uri) {
      15 |       // reference URI, but no actual payload to host there!
    > 16 |       throw Error(SIOPErrors.REFERENCE_URI_NO_PAYLOAD)
         |             ^
      17 |     } else if (checkRequestObject) {
      18 |       throw Error(SIOPErrors.BAD_PARAMS)
      19 |     }


There is a payload in the opts root, but that doesn't end up in the requestObject

File: packages/siop-oid4vp/lib/__tests__/AuthenticationRequest.request.spec.ts (hunk 2 above)
Comment (sanderPostma) [reply packages/siop-oid4vp/lib/__tests__/AuthenticationRequest.request.spec.ts]: Actually test "succeed when requesting with a valid PD" fails when I do not provide the claims on line 553
> TypeError: Cannot read properties of undefined (reading 'vp_token')

File: packages/siop-oid4vp/lib/authorization-response/PresentationExchange.ts (hunk 6 above)
Comment (sanderPostma) [reply packages/siop-oid4vp/lib/authorization-response/PresentationExchange.ts]:
```
    // When there are MDoc credentials among the selected ones, filter those out as pex does not support mdoc credentials
    const filteredCredentials = this.removeMDocCredentials(selectedCredentials) 
    return await this.pex.verifiablePresentationFrom(presentationDefinition, filteredCredentials, presentationSignCallback, signOptions)
```

File: packages/siop-oid4vp/lib/rp/Opts.ts (hunk 13 above)
Comment (sanderPostma) [reply packages/siop-oid4vp/lib/rp/Opts.ts:63]: Guess we'll never find out anymore...
image

I left the comment there in case anyone else would decide to "fix" something in the future

File: packages/siop-oid4vp/lib/rp/RP.ts (hunk 16 above)
Comment (sanderPostma) [reply packages/siop-oid4vp/lib/rp/RP.ts:234]: fixed
